File: Back_end/src/main/java/com/bilel/SpringBoot_TP01/services/UserService.java
package com.bilel.SpringBoot_TP01.services;

import com.bilel.SpringBoot_TP01.entities.Role;
import com.bilel.SpringBoot_TP01.entities.User;

import java.util.List;

public interface UserService {
    User saveUser(User user);
    User updateUser(User user);
    User findByUsername(String username);
    User findById(Long userId);
    User grantRoleToUser(User user, Role role);
    void revokeRoleFromUser(User user, Role role);
    List<User> findAllUsers();

    void deleteUser(Long userId);
}

File: Front_End/src/app/services/user/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { User } from 'src/app/models/User.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiServerUrl = `${environment.apiBaseUrl}/api/users`;

  constructor(private http: HttpClient) { }

  public getUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiServerUrl}/all`);
  }

  public getUser(userId: number): Observable<User> {
    return this.http.get<User>(`${this.apiServerUrl}/get/${userId}`);
  }

  public addUser(user: User): Observable<User> {
    return this.http.post<User>(`${this.apiServerUrl}/create`, user);
  }

  public grantRoleToUser(userId: number, roleId: number): Observable<User> {
    return this.http.post<User>(`${this.apiServerUrl}/grant/role/${roleId}/user/${userId}`, null);
  }

  public revokeRoleFromUser(userId: number, roleId: number): Observable<void> {
    return this.http.post<void>(`${this.apiServerUrl}/revoke/role/${roleId}/user/${userId}`, null);
  }

  public updateUser(user: User, userId: number): Observable<User> {
    return this.http.patch<User>(`${this.apiServerUrl}/update/${userId}`, user);
  }

  public deleteUser(userId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiServerUrl}/delete/${userId}`);
  }
}
